### PR TITLE
Fixed stream reset and peer is not directly connected errors

### PR DESCRIPTION
**Pull request by iulianpascalau**
Direct sender fixed, stream reset issue should not appear
TopicResolverSender tries to send the request to the next peer if a send to a peer fails.

### CODE REVIEW

File: dataRetriever/resolvers/topicResolverSender/topicResolverSender.go [hunk 1]
```diff
@@ -66,37 +63,58 @@ func (trs *topicResolverSender) SendOnRequestTopic(rd *dataRetriever.RequestData
...
+	msgSentCounter := 0
+	for idx := range shuffledIndexes {
+		peer := diffList[idx]
+
 		err = trs.messenger.SendToConnectedPeer(topicToSendRequest, buff, peer)
-		if err != nil {
-			log.Debug(err.Error())
-		} else {
-			messageSent = true
+		if err == nil {
```
Comment (sasurobert) [inline dataRetriever/resolvers/topicResolverSender/topicResolverSender.go]: change to if err != nil -> continue

File: p2p/libp2p/directSender.go [hunk 2]
```diff
@@ -43,21 +47,24 @@ func NewDirectSender(
...
 	}
-
 	if ctx == nil {
 		return nil, p2p.ErrNilContext
 	}
-
 	if messageHandler == nil {
 		return nil, p2p.ErrNilDirectSendMessageHandler
 	}
+	mutexes, err := NewMutexHolder(maxMutexes)
```
Comment (sasurobert) [inline p2p/libp2p/directSender.go]: add an empty line before.

File: p2p/libp2p/directSender.go [hunk 3]
```diff
@@ -145,6 +149,10 @@ func (ds *directSender) Send(topic string, buff []byte, peer p2p.PeerID) error {
 		return p2p.ErrMessageTooLarge
 	}
 
+	mut := ds.mutexes.Get(string(peer))
```
Comment (sasurobert) [inline p2p/libp2p/directSender.go]: rename suggestion for variable: mutexForPeer ?

File: p2p/libp2p/mutexHolder.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/storage/lrucache"
+)
+
+// MutexHolder holds a cache of mutexes: pairs of (key, *sync.Mutex)
+type MutexHolder struct {
+	//generalMutex is used to assure access to the already concurrent safe lrucache is being serialized
```
Comment (sasurobert) [inline p2p/libp2p/mutexHolder.go]: //generalMutex is used to serialize the access to the already concurrent safe lrucache

File: p2p/libp2p/mutexHolder.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	generalMutex sync.Mutex
```
Comment (sasurobert) [inline p2p/libp2p/mutexHolder.go]: is it a must to protect this ?

File: dataRetriever/resolvers/topicResolverSender/topicResolverSender.go (hunk 1 above)
Comment (iulianpascalau) [reply dataRetriever/resolvers/topicResolverSender/topicResolverSender.go]: done

File: p2p/libp2p/directSender.go (hunk 2 above)
Comment (iulianpascalau) [reply p2p/libp2p/directSender.go]: done

File: p2p/libp2p/directSender.go (hunk 3 above)
Comment (iulianpascalau) [reply p2p/libp2p/directSender.go]: done

File: p2p/libp2p/mutexHolder.go (hunk 4 above)
Comment (iulianpascalau) [reply p2p/libp2p/mutexHolder.go]: changed

File: p2p/libp2p/mutexHolder.go (hunk 5 above)
Comment (iulianpascalau) [reply p2p/libp2p/mutexHolder.go]: yes, there is a get and eventually some put's there that must be made atomically. We do not want a go routine to set a mutex in a lru on a key that had an already set mutex there.

**Pull request merged**